### ISSUE TITLE
Add 'singleton' flag to field mappers?

### ISSUE THREAD

**Issue by jtibshirani** [elastic/elasticsearch#58523]
Datastreams have a first-class concept of a timestamp field. Each document in the datastream must contain exactly one value for the designated timestamp field, so that we know where to route the document when partitioning by time. In #58582 we're adding index-time validation for this requirement. The current implementation is very narrow in scope and adds special document parsing logic just for datastream timestamps.

I was wondering if this validation could be useful more generally. We could add a 'singleton' flag to field mappers -- when it is set, the mapper will verify that it encounters one (and only one) value in each document:

```
"product_identifier": {
  "type": "keyword",
  "singleton": true
}
```

Such an option could be helpful for modeling fields that an application requires to have exactly one value, like identifiers, timestamps, content types, etc.

**Comment by webmat** [comment]
I'd like to clarify one part:

> Each document in the datastream must contain exactly one timestamp

I assume this sentence applies only to the field where `singleton` is set. In your example, `@timestamp` contains exactly one timestamp, and not an array of timestamps. This I'm totally on board with 👍 

However the way this is currently worded makes it sound like no other date fields would be allowed (e.g. an event with `@timestamp`, `event.start` and `event.end` populated). I assume this is not the case here?

Stepping out of the timestamp example, I like this feature in general, not just for datastreams. Elasticsearch has long been flexible with fields containing either a single value or an array. However since ECS came out, we've gradually been clarifying which ECS fields are expected to contain an array, in order to make the events more predictable for consumers of the data. 

This new addition would help clarify 3 acceptable formats about a field:

* it *must* be a single value
* it *must* be an array of values
* it's (still) unspecified (default Elasticsearch behaviour)

**Comment by jtibshirani** [comment]
> However the way this is currently worded makes it sound like no other date fields would be allowed (e.g. an event with @timestamp, event.start and event.end populated). I assume this is not the case here?

Your assumption is right, I've updated the description to clarify the wording.

**Comment by ruflin** [comment]
If I understand it correctly, `constant_keyword` today is already a singleton field? In general ++ on this feature?

**Comment by jtibshirani** [comment]
> If I understand it correctly, constant_keyword today is already a singleton field?

Not exactly, because documents are allowed to not specify a value for `constant_keyword`. When performing a search, we know to fill in the missing field with the constant value from the mappings. But you're right in that they appear like singleton fields from a search API perspective.

**Comment by romseygeek** [comment]
We discussed this in the search meeting and agree that it would be a useful addition to at least some of the field mappers.  We think that `allow_multi_fields:true|false` might be a better parameter name to use as it's more obvious that we're not enforcing that a single value is present.

**Comment by webmat** [comment]
From the start I've interpreted this as a way to ensure a field would not contain an array of values. E.g. preventing the following:

```JSON
{ "@timestamp": [ "1597425324", "1597425325" ] }
```

And enforcing that a field has a single value, e.g.:

```JSON
{ "@timestamp": "1597425324" }
```

My understanding is that this feature is unrelated to whether or not a field has additional multi-fields.

**Comment by mayya-sharipova** [comment]
I think @romseygeek meant `allow_multiple_values`:
- when set `true`, we will allow multiple values
- when set `false`, we will not allow multiple values.  A value may be a single value, or it may not exist. 

Or is the requirement with `singleton` is that every document **must** has this field and only a single value?

**Comment by jtibshirani** [comment]
@romseygeek @mayya-sharipova yes, the `singleton` flag would ensure there is exactly one value, not just 'at most' one value (as mentioned in the description). This is part of the usefulness of the feature -- for example with time-based data, we may want to ensure that every document has a timestamp and can be associated with a bucket based on time.

**Comment by astefan** [comment]
@markharwood agree on dealing with the field caps API potential additional in a different issue. I don't think QL has any use in a mapping level flag, as we do not look at the indices mappings.

**Comment by markharwood** [comment]
> don't think QL has any use in a mapping level flag, as we do not look at the indices mappings.

@astefan I don't expect you would to need to look at mappings to benefit from this. We'd surface the mapping level flags in field cap summaries.

When it come to cardinalities I'm leaning towards the policy of **_enforce-via-mapping_** rather than  **_report-on-index-state_** because:
1) Costs of deriving field-caps information are higher when you have to poke around all segments of all indices to reverse-engineer actual cardinality info.
2) Field caps would require frequent calls as cardinality info is not stable - a field reported as single-valued one instant could turn to a multi valued one the next with the addition of one more document. 
3) Reverse-engineering cardinalities from poking around Lucene indices will not work with all field types. A single-valued JSON property could be indexed any number of ways by a field - as a single token or many or not at all in the case of runtime fields.

**Comment by markharwood** [comment]
When it comes to enforcement I think enforcing a maximum cardinality will be much cheaper/more viable than enforcing a minimum cardinality.
Checking the max cardinality is simple because there's one central piece of logic in the DocumentParser class that parses JSON arrays. We can put a simple field-specific limit in there.
However, checking that fields are present in every doc i.e. when min_cardinality is 1 involves more computation for each doc:
1) All fields in a mapping with min_cardinality >0 have to be loaded into a "required_fields" set
2) Each field encountered in the JSON removes any corresponding name from the "required_fields" set
3) An error is thrown if the required_fields set is not empty for a fully parsed document
4) Extra complexity occurs within objects or nested fields that are allowed multiples but their leaf fields are not.

I'm concerned that this min_cardinality checking would add overhead to one of the hottest loops - the indexing logic.
For this reason I propose that we only offer an `allow_multiple_values` boolean for fields rather than offer both a min and max cardinality validation.

**Comment by jpountz** [comment]
One benefit of enforcing in the mappings that I can think of is that we might be able to speed up indexing for single-valued fields by using `NumericDocValuesField` instead of `SortedNumericDocValuesField` and `SortedDocValuesField` instead of `SortedSetDocValuesField`. It might also make it easier to reuse Field instances, which is another factor that helps speed up indexing. I'm unsure about the sort of speedup that we should expect though.

> Costs of deriving field-caps information are higher when you have to poke around all segments of all indices to reverse-engineer actual cardinality info

FWIW we already have such logic when an `index_filter` is provided, since we need to examine all segments to know if any of them intersects with the query.

> (perhaps even more so on frozen tier)

Index statistics are loaded in memory when opening a shard, so it wouldn't be slower on the frozen tier. However nodes on the frozen tier are denser, so if they would have more shards to check than hot nodes.

The worry I have about the "enforce-via-mapping" approach is adoption:
 - it wouldn't work on indices created before this feature is introduced,
 - if you forgot to set it at index creation time, you cannot add it later without reindexing,
 - some users (e.g. Logging use-cases) might be nervous about using this feature because they'd rather index documents that are multi-valued even though they shouldn't than reject them.

Furthermore the "enforce-via-mapping" approach brings unique challenges too. For instance if you have 100 data streams and one of them doesn't set the `allow_multiple_values` flag on the `agent.id` field even though it is single-valued, then calling field caps on an index pattern that matches these 100 data streams will report `agent.id` as being multi-valued.

To me the strongest argument against the "report-on-index-state" approach is the performance overhead. Maybe we could build a prototype and try to evaluate the performance impact to see how bad it actually is?

**Comment by markharwood** [comment]
>we might be able to speed up indexing for single-valued fields by using NumericDocValuesField instead of SortedNumericDocValuesField and SortedDocValuesField instead of SortedSetDocValuesField

I tried this out for `keyword` and `date` fields for a couple of smallish datasets but didn't see any noticeable uplift to indexing speeds. Haven't tried recycling `Field` instances yet.

**Comment by dakrone** [comment]
Was this ever decided on? It seems like it would be a useful feature.

**Comment by ruflin** [comment]
This popped up again also in the context of the routing processor that now landed: https://github.com/elastic/elasticsearch/pull/76511 As soon as a value is multi field, we might not be able to route the document. At the same time the fields that are likely to be used for routing should never be multi field in the first place. It would be nice to even enforce this.

Our need for a singleton field is not driven by any performance improvements but much rather strictness on the data. With https://github.com/elastic/elasticsearch/pull/95329 we become more lenient on what data we accepts for `logs-*-*`. It means even if it is a singleton field, it would still be accepted just not indexed.

Could we start this effort by having the option on keyword fields? It seems @markharwood made some tests quite some time ago on this which means the code was changed for it. Does anyone know if this code is still around somewhere?

**Comment by abatsakis** [comment]
We can discuss the feature independently, but why we don't want a routing field to ever be multi-valued? Also, are you gonna force the rerouting processor to work only on fields with the singleton entry?

**Comment by ruflin** [comment]
The discussion on the reroute processor happened here: https://github.com/elastic/elasticsearch/pull/76511#discussion_r1164154137 For now, routing only works on fields with a singleton entry and otherwise throws and exception. The problem with multi value is, which value do we pick for routing? @felixbarny

**Comment by javanna** [comment]
This is getting some traction again recently, I recently discussed it with @nik9000 as well in the context of esql. We discussed it today again within the Search team and we said it's a thing that we would like to do once some of us has time.

**Comment by abatsakis** [comment]
@ruflin the answer to your question is... it doesn't matter :) If a user doesn't explicitly handle multi-valued fields in their routing processors, then the data are most likely bogus so you can do whatever you want with them, e.g. route them to the error store explicitly, drop them, pick one value randomly or whatever.

This is something that can happen anyway because someone may try to route a field that is not designated as singleton, unless you will require routing rules to be applied only to those fields (which I don't think you want to). Same for ESQL.

In other words, I don't see good reasons for this feature. Maybe I am missing something though, so before you intend to pick it up, let's talk :)

**Comment by javanna** [comment]
Some clarifications:

> I get the following error: "reason": "[11:1] failed to parse: data stream timestamp field [@timestamp] encountered multiple values". Seems like we already have a singleton field here (which is good).

Yes, we perform this check specifically for the `@timestamp` field. Quoting the description of this issue "Datastreams have a first-class concept of a timestamp field. Each document in the datastream must contain exactly one value for the designated timestamp field, so that we know where to route the document when partitioning by time.". This issue was originally to generalize this check and potentially apply it to other fields. I see that for a big part of the discussion  though we actually discussed exposing whether a field has multiple values or not, which is a different problem, meaning adding a flag that specifies whether a field is allowed to have multiple values or not, but that does not mean each document must have a value for it. 

While we are currently discussing the possibility of adding the `allow_multiple_values` flag which is covered by #80825, we have no concrete plans on adding the `singleton` flag. We acknowledged that it may help but we don't have much data and concrete scenarios where that would help, which is pretty much the reason why we have not implemented it so far: it has not been a priority. For this reason I am closing this issue, we will reserve to reopen it if we collect feedback on concrete scenarios where adding a singleton flag would help.
